add reproducibility and source tracing documentation

deitch opened the pull request:
We have had several requests describing how EVE is reproducible. How can we reliably reproduce the same bits over and over again. This is good practice, but also important for debugging, auditing and security concerns. I call this "downward" - given the source, how can I get to (functionally) the same bits over and over again. This is especially relevant in light of OCI registry (like Docker Hub and quay.io) mutable tags, and Alpine apk package changes.

The `REPRODUCIBILITY.md` document here addresses that question.

Additionally, we have had requests describing how to trace all of the sources in EVE. Given an EVE OS image, how can we reliably trace the precise software used to create it? This is good practice, but also important for debugging, auditing and security concerns. It also may become important when managing various open source licenses. I call this "upward" - given the binary, how can I get to the sources.

The `SOURCES.md` document here addresses that question.

zedi-pramodh reviewed (commented):
If the roofs.yml is generated during build for template images ? Is that generated file committed somewhere ? Else how can we be sure of the hashes used for template images ?

deitch commented:
> If the roofs.yml is generated during build for template images ? Is that generated file committed somewhere ? Else how can we be sure of the hashes used for template images ?

No, it isn't committed. However, the template processing uses the calculated hashes from the package directories to generate the template file.

In the development process, people often change one or several packages, and then want to build eve right away. This is why it is templated, so that a call to `make` can generate it on the fly.

I had spent some time a few years back looking at eliminating the templating, but didn't have the time to go all the way through it. My initial thoughts were to have actual tags in the `rootfs.yml`, no templates. When you are done changing, you update the file with the latest commits. For development, you would have some way to generate an external file, not committed to git, so maybe `make eve` and `make eve DEV=true` or similar.

zedi-pramodh commented on docs/REPRODUCIBILITY.md:
````diff
@@ -0,0 +1,267 @@
+lfedge/eve-pillar:b20146debd08bbfffb23f53fc9f5c3c4af377f67
+```
+
+With the template complete, every OCI container image used to compose an EVE operating system is in the generated `rootfs.yml`, each of which
````
Not clear on how we get back to this generated roofs.yml in future to find the exact hashes used in the templated images ?

deitch replied:
I will update the doc to show that this is the purpose of it.

deitch replied:
Take a look at the introduction again.

zedi-pramodh commented on docs/REPRODUCIBILITY.md:
```diff
@@ -0,0 +1,267 @@
+All of the packages used in EVE or sourced from one of two locations:
+
+* `docker.io/linuxkit/*` whose sources are [https://github.com/linuxkit/linuxkit/tree/master/pkg/](https://github.com/linuxkit/linuxkit/tree/master/pkg/)
+* `docker.io/lfedge/eve-*` whose sources are in this repository under [../pkg](../pkg)
```
Who puts those here ?As part of build process ?

deitch replied:
Yeah, I probably should clarify that.

deitch replied:
Take a look again

zedi-pramodh commented on docs/REPRODUCIBILITY.md:
```diff
@@ -0,0 +1,267 @@
+
+The usage of `PKGS` environment variable and the script `eve-alpine-deploy.sh` is a simplified way to install packages. It is the equivalent of calling:
+`apk add alpine-baselayout musl-utils bash glib squashfs-tools util-linux e2fsprogs e2fsprogs-extra keyutils dosfstools coreutils sgdisk smartmontools`, but
+do so from the cache.
```
Ok so since the base image is always fixed to a hash, these packages installed are guaranteed to be same every time right ?

deitch replied:
Yes. As long as you use the same hash-based version of the base image, then `apk add` will use the fixed ones in the cache, which is unchanged.

zedi-pramodh commented:
I am not concerned about dev builds. But for a released build say 8.5.3 how do we know the hashes of templated images shipped in the image ? I know we can reliably get those by running show-tag on the 8.5.3 tagged directory, but how can I prove those are the hashes used in the eve-image ?

deitch commented:
> I am not concerned about dev builds.

I am. 😄 Solving for that will make it easier to get past templates.

> But for a released build say 8.5.3 how do we know the hashes of templated images shipped in the image ? I know we can reliably get those by running show-tag on the 8.5.3 tagged directory, but how can I prove those are the hashes used in the eve-image ?

That is the bin2src question, is it not? That will be answered in a different file (to be added to this PR, currently work in progress). The src2bin question is, "given this commit that I have checked out, how do I know that building will produce functionally identical binaries?"

zedi-pramodh commented:
yes its bin2src will wait for your updated commit.

zedi-pramodh reviewed (commented):
One nit point. I would rename the file to EVE-IMAGE-REPRODUCABILITY.md, just reproducibility does not mean much unless you actually know what it means :)

deitch commented:
Good point. Renamed.

deitch commented:
Yetus actually is complaining that I used fenced code blocks instead of indented. Not sure who decided that was a valid rule; I am ignoring it.

eriknordmark commented on docs/EVE-IMAGE-REPRODUCIBILITY.md:
```diff
@@ -0,0 +1,274 @@
+
+We write "functionally identical" rather than "precisely identical" or just "identical", because the compilation and composition process can change
+elements like timestamps or randomly generated data, leading to minor differences. Ideally, two builds of EVE from the same source would be
+precisely identical; that is a future goal. However, functionally identical images are a requirement for EVE.
```
Do we know whether we have such issues today?
It would make sense to try a make pkgs rootfs and see if the shas differ.
If they do it will be hard for a user to verify that something tagged with x.y.z built by lf-edge (or built by anybody else) was really built from the x.y.z source since they can't redo the build on their system and compare the shas.

deitch replied:
I don't know. I haven't tried it. I also thought about trying it out, but there were different possible scenarios. I would prefer if we can release this document (when ready) with this caveat, and then we can put a targeted effort into either verifying that it is not an issue or fixing it.

eriknordmark replied:
OK

eriknordmark commented on docs/EVE-IMAGE-REPRODUCIBILITY.md:
```diff
@@ -0,0 +1,274 @@
+
+In order to resolve this issue, both the external/upstream `linuxkit/*` packages and the `lfedge/eve-*` packages in this repository follow the same principles.
+
+1. Create an alpine "base" image: `linuxkit/alpine` and `lfedge/eve-alpine`
```
Given that we have lfedge/eve-alpine-base as well, should we call this the alpine "cache" image? We informally refer to this as a cache.

deitch replied:
We do have a bit of a naming overload issue, do we not?

I can rename most of these to "cache". My concern was that a cache implies something "on the side", that I choose to use, where these are `FROM`, i.e. something I build on top of. It isn't that my downstream package builds on `eve-alpine-base` and somewhere configures "use this other image as my cache". Rather, it is that downstream uses `eve-alpine` _as its base_ via `FROM lfedge/eve-alpine`.

What do you think?

deitch replied:
I updated it, I think this is clearer.

eriknordmark commented on docs/EVE-IMAGE-REPRODUCIBILITY.md:
````diff
@@ -0,0 +1,274 @@
+
+```dockerfile
+FROM lfedge/eve-alpine:145f062a40639b6c65efa36bed1c5614b873be52 as build
+ENV PKGS alpine-baselayout musl-utils bash glib squashfs-tools util-linux e2fsprogs e2fsprogs-extra keyutils dosfstools coreutils sgdisk smartmontools
````
storage-init does not have a separate BUILD_PKGS. Would it be better to use an example which does?

deitch replied:
changed to use `dom0-ztools` which has both, and added an explanation.

eriknordmark commented on docs/EVE-IMAGE-REPRODUCIBILITY.md:
```diff
@@ -0,0 +1,274 @@
+
+This document describes the reproducibility structures and process of EVE. It answers the "source-to-binary" question: 
+
+    given that I have a specific version of this source code repository, how does the process ensure that I will get an
```
I think the statement is really "either produce a functionally eqivalent or fail".
For example, a curl/wget of a particular hashed piece of code might fail to find that if I try rebuilding the same eve tag a year from now. Don't know how likely it is that other parts disappear; the alpine caching will handle alpine pkg versions which are removed from alpine.

Don't know if you want to mention this "or fail" in this document.

deitch replied:
I actually do, that is a very good point. Getting it now.

deitch replied:
Done

deitch commented:
Added the sources document; will review the outstanding comments later.

zedi-pramodh commented:
I don't see sources.md in the commit.

deitch commented:
> I don't see sources.md in the commit.

it is odd. I see it on my local, I see the commit, I see the push. I will try again.

deitch commented:
Ah, now it is there @zedi-pramodh 🤷‍♂️

eriknordmark commented on docs/EVE-IMAGE-SOURCES.md:
```diff
@@ -0,0 +1,616 @@
+1. Binary to source commit
+1. Binary to builder configuration
+1. Container images to container sources
+1. Container ources to contents:
```
typo "ources"

deitch replied:
Never heard of "ources"? It is an old French variant on "bears". 😆

eriknordmark commented on docs/EVE-IMAGE-SOURCES.md:
````diff
@@ -0,0 +1,616 @@
+
+* For Docker images, look in `/bits/eve_version`. For example:
+```
+$ docker run -it --entrypoint=sh lfedge/eve:0.0.0-master-0c6de671-kvm -c "cat /bits/eve_version"
````
docker run lfedge/eve:0.0.0-master-0c6de671-kvm version
is less typing.

deitch replied:
I forgot that you can do that! Updating.

eriknordmark commented on docs/EVE-IMAGE-SOURCES.md:
```diff
@@ -0,0 +1,616 @@
+
+## Binary to Builder Configuration
+
+The specific configuration used to build this distribution of EVE is distriubted inside the EVE OS image.
```
typo "distriubted"

deitch replied:
Thanks, fixed.

eriknordmark commented on docs/EVE-IMAGE-SOURCES.md:
```diff
@@ -0,0 +1,616 @@
+The git commit for the above source is given in the label `"org.opencontainers.image.revision": "12a6fb8d29a6b47f998cd077dfb92213f0e6a55f"`, i.e.
+you can get the same commit via `git checkout 12a6fb8d29a6b47f998cd077dfb92213f0e6a55f`.
+
+The above _should_ be the same as the current git commit.
```
What does the _should_ imply? And why the reference to "current"? Did you mean what was current when it was built?

deitch replied:
No, I meant, the one that currently was checked out in the earlier line. I will clarify it.

deitch replied:
Updated, look when you have a chance, please.

eriknordmark commented on docs/EVE-IMAGE-SOURCES.md:
```diff
@@ -0,0 +1,616 @@
+or files not saved in git. If there are, it treats the repository as "dirty", and appends both `dirty` and a hash of all of the file
+contents in the directory to the tag. Thus `lfedge/eve-pillar:17837a9fcd05c765e9a1f6707b2e48f0f1dd215b-amd64` means that it checked,
+and there were no uncommitted changes or files in the source directory. If there were, it would have indicated on the tag as
+`lfedge/eve-pillar:17837a9fcd05c765e9a1f6707b2e48f0f1dd215b-dirty-<hash_of_file_contents>-amd64`.
```
I think one can make the statement that if -dirty is included it is not possible to know the complete source, since some files very modified locally in the workspace where the image was built. Should we clarify that in the document?

deitch replied:
Yeah, good point. You _might_ get lucky and have the same files - if the hash works out, i.e. `lkt pkg show-tag` gives the same result - but you might not. There is no way to "get to it", like `git checkout`. Will add.

eriknordmark commented on docs/EVE-IMAGE-SOURCES.md:
```diff
@@ -0,0 +1,616 @@
+
+### Go modules
+
+Most of the compiled content used in EVE is written in [go](http://golang.org). For all go package dependencies, all sources
```
We compile the Linux kernel etc as well.
So maybe instead of "compiled content" we should say "code unique to EVE"?

deitch replied:
Sure, I am good with that.

eriknordmark commented on docs/EVE-IMAGE-SOURCES.md:
```diff
@@ -0,0 +1,616 @@
+
+Some of the above include only semver, while one includes the precise git commit and date or commit.
+
+Further, go modules preserve the actual hashes in [go.sum](../pkg/pillar/go.sum), including those for which semver was used.
```
It might be useful to add something about how to take a line in go.sum and find the actual commit of the source code used.

deitch replied:
OK, added some more. Have a look.

eriknordmark replied:
Looks good to me. I don't know if this is sufficient to create a complete list of URLs to all source. @ruslan-zededa are you looking at that? We can chat offline.

eriknordmark commented on docs/EVE-IMAGE-REPRODUCIBILITY.md:
```diff
@@ -0,0 +1,286 @@
+# EVE Reproducibility
```
It probably makes sense to put a link in BUILD.md to this document.

deitch replied:
Good idea. Adding in.

eriknordmark commented on docs/EVE-IMAGE-REPRODUCIBILITY.md:
```diff
@@ -0,0 +1,286 @@
+
+Further, the caches maintain hashes of the local cache contents, to ensure that any changes are known and captured.
+
+To make this more concrete, we will walk through a specific example, `lfedge/eve-dom0-ztools`, which is based in [../pkg/dom0-ztools](../pkg/dom-ztools/).
```
Bad link; dom-ztools should be dom0-ztools

deitch replied:
Oops. Fixing.

eriknordmark commented on docs/EVE-IMAGE-REPRODUCIBILITY.md:
````diff
@@ -0,0 +1,286 @@
+linuxkit/alpine:c9b0f6a435b663b98952d67f4c6f105c310d0a21-amd64
+```
+
+and the file [versions.x86_64](https://github.com/linuxkit/linuxkit/tree/master/tools/alpine/versions/x86_64) begins with:
````
bad link.

deitch replied:
Oops, sorry. Fixing.

deitch commented:
I updated to simplify the source tracing for Alpine packages. I avoided it on the first run, thinking it would be easier to go to the source apk in our cache images rather than read the concise `installed` database file, but realized that it was more work for the user that way. Updated.

deitch commented on docs/EVE-IMAGE-REPRODUCIBILITY.md:
````diff
@@ -0,0 +1,288 @@
+The calculation is provided by running `linuxkit pkg show-tag dir/`. For example:
+
+```sh
+$ linuxkit pkg show-tag pkg/pillar
````
I think that is too much. We describe it in `BUILD.md`, and we added a link in this doc, thanks to @eriknordmark's suggestion.

deitch commented on docs/EVE-IMAGE-SOURCES.md:
```diff
@@ -0,0 +1,729 @@
+## Binary to Builder Configuration
+
+The specific configuration used to build this distribution of EVE is distributed inside the EVE OS image.
+The actual configuration is available in binary `rootfs.img`, specifically in `/etc/linuxkit-eve-config.yaml`. For example:
```
Sure, we can add that.

deitch replied:
Updated

deitch commented on docs/EVE-IMAGE-REPRODUCIBILITY.md:
````diff
@@ -0,0 +1,288 @@
+apk add bash=5.1.16-r2
+```
+
+The above does not completely solve it, as sometimes alpine packages change their bits _even for the same version_. Thus, running
+`apk add bash=5.1.16-r2` on the same base `alpine:3.16` can give different bits on different days. Alpine packages are changed "under the covers"
+when bugs are found or updates are provided.
````
Updated.

eriknordmark reviewed (approved):
LGTM

eriknordmark commented:
@deitch There are three yetus complaints (out of 10) which are not about the fenced vs intended. Shouldn't those be fixed?

deitch commented:
> There are three yetus complaints (out of 10) which are not about the fenced vs intended. Shouldn't those be fixed?

10 yetus complaints now:

* 1 about space at end of line (good call, I just fixed that now)
* 7 about indented vs fenced - we _really_ should remove that from yetus, as it is just wrong. Fenced is 100% valid.
* 2 about fenced requiring a language - I am ignoring this one as well. Sometimes you just want something to be highlighted in a block without it being a specific language.

I fixed the one that should be fixed; we can ignore the rest.
